### PR TITLE
API proxy struct codegen

### PR DESCRIPTION
**Pull request by magik6k**
Includes https://github.com/filecoin-project/lotus/pull/5844

The main part is just 300LOC in ` gen/api/proxygen.go`, rest of the diff is moving permissions and codegen

Like docsgen, for now it's hardcoded for the `api/` package, that will get fixed when porting to the versioned api

### CODE REVIEW

File: api/api_common.go [hunk 1]
```diff
@@ -19,60 +19,60 @@ type Common interface {
 
 	// MethodGroup: Auth
 
-	AuthVerify(ctx context.Context, token string) ([]auth.Permission, error)
-	AuthNew(ctx context.Context, perms []auth.Permission) ([]byte, error)
+	AuthVerify(ctx context.Context, token string) ([]auth.Permission, error) // perm:read
```
Comment (Kubuxu) [inline api/api_common.go:22]: directives, in contrast to comments,  are written without leading space.

File: gen/api/proxygen.go [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+{{end}}
+
+`)
+	return err
+}
+
+func doTemplate(w io.Writer, info interface{}, templ string) error {
+	t := template.Must(template.New("").
+		Funcs(template.FuncMap{
+			"ReadHeader": func(rdr string) string {
```
Comment (Kubuxu) [inline gen/api/proxygen.go:294]: Unused.

**Comment by magik6k** [comment]
Ah, openrpc was getting param names from the struct, which is why those files are now smaller - diff from master:
```
10c10
<       "description": "```go\nfunc (c *FullNodeStruct) BeaconGetEntry(ctx context.Context, epoch abi.ChainEpoch) (*types.BeaconEntry, error) {\n\treturn c.Internal.BeaconGetEntry(ctx, epoch)\n}\n```",
---
>       "description": "```go\nfunc (s *FullNodeStruct) BeaconGetEntry(p0 context.Context, p1 abi.ChainEpoch) (*types.BeaconEntry, error) {\n\treturn s.Internal.BeaconGetEntry(p0, p1)\n}\n```",
15c15
<           "name": "epoch",
---
>           "name": "p1",
66c66
<         "url": "https://github.com/filecoin-project/lotus/blob/master/api/apistruct/struct.go#L893"
---
>         "url": "https://github.com/filecoin-project/lotus/blob/master/api/apistruct/struct.go#L885"

```

File: Makefile [hunk 3]
```diff
@@ -325,7 +325,11 @@ type-gen:
 method-gen:
 	(cd ./lotuspond/front/src/chain && go run ./methodgen.go)
 
-gen: type-gen method-gen
+api-gen:
+	go run ./gen/api > api/apistruct/struct.go
+	goimports -w api/apistruct
+	goimports -w api/apistruct
```
Comment (vyzo) [inline Makefile:331]: why twice?

File: gen/api/proxygen.go [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+	case *ast.ChanType:
+		subt, err := typeName(t.Value)
+		if err != nil {
+			return "", err
+		}
+		if t.Dir == ast.SEND {
+			subt = "->chan " + subt
+		} else {
+			subt = "<-chan " + subt
+		}
```
Comment (vyzo) [inline gen/api/proxygen.go:112]: how do we handle bidirectional channels? Or are they just not a thing in the api?

**Review by vyzo — COMMENTED** [review]
this looks good to me and it's definitely an improvement over having to do it by hand.

I suppose the v1 scaffolding will come at a later pr or is it planned for here?

File: Makefile (hunk 3 above)
Comment (magik6k) [reply Makefile:331]: Apparently calling it once won't fully fix imports

File: gen/api/proxygen.go (hunk 4 above)
Comment (magik6k) [reply gen/api/proxygen.go:112]: Those don't work over jsonrpc

**Comment by Kubuxu** [comment]
It was also my mistake re binary files, I through the 99% meant 99% smaller, it probably means 99% different.

**Review by vyzo — APPROVED** [review]

**Review by Kubuxu — APPROVED** [review]
Let's squash as it doesn't build in the middle.

**Pull request merged**